homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,242 @@
+        """Return a coil with data and check for validity."""
+        if coil := self.get_coil_data(coil):
+            return coil.value
+        return None
+
+    def get_coil_float(self, coil: Coil | None) -> float | None:
+        """Return a coil with float and check for validity."""
+        if value := self.get_coil_value(coil):
+            return float(value)
+        return None
+
+    async def async_write_coil(
+        self, coil: Coil | None, value: int | float | str
+    ) -> None:
+        """Write coil and update state."""
+        if not coil:
+            raise Exception("No coil available")
+
+        coil.value = value
+        coil = await self.connection.write_coil(coil)
+
+        if self.data:
+            self.data[coil.address] = coil
+            self.async_update_listeners()
+
+    async def _async_update_data(self) -> dict[int, Coil]:
+        @retry([0.0, 0.5], (CoilReadException))
+        async def read_coil(coil: Coil):
+            return await self.connection.read_coil(coil)
+
+        callbacks: dict[int, list[CALLBACK_TYPE]] = {}
+        for update_callback, context in list(self._listeners.values()):
+            assert isinstance(context, set)
+            for address in context:
+                callbacks.setdefault(address, []).append(update_callback)
+
+        result: dict[int, Coil] = {}
+
+        if self.data is None:
+            self.data = {}
+
+        for address, callback_list in callbacks.items():
+            try:
+                coil = self.heatpump.get_coil_by_address(address)
+                self.data[coil.address] = result[coil.address] = await read_coil(coil)
+            except (CoilReadException, TooManyTriesException) as exception:
+                self.logger.warning("Failed to update: %s", exception)
+            except CoilNotFoundException as exception:
+                self.logger.debug("Skipping missing coil: %s", exception)
+
+            for update_callback in callback_list:
+                update_callback()
+
+        return result
+
+
+class CoilEntity(CoordinatorEntity[Coordinator]):
+    """Base for coil based entities."""
+
+    _attr_has_entity_name = True
+    _attr_entity_registry_enabled_default = False
+
+    def __init__(
+        self, coordinator: Coordinator, coil: Coil, entity_format: str
+    ) -> None:
+        """Initialize base entity."""
+        super().__init__(coordinator, {coil.address})
+        self.entity_id = async_generate_entity_id(
+            entity_format, coil.name, hass=coordinator.hass
+        )
+        self._attr_name = coil.title
+        self._attr_unique_id = f"{coordinator.unique_id}-{coil.address}"
+        self._attr_device_info = coordinator.device_info
+        self._coil = coil
+
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        return self.coordinator.last_update_success and self._coil.address in (
+            self.coordinator.data or {}
```

Comment:
What does the `or {}` do for you? Is it not enough to just check for the coil address in the coordinator data?

Reply:
data can be None before first refresh.

Reply:
Can we set it to empty dict in `__init__`?

Reply:
Yup